From b9dd5ca1c1758c9eef87cc4e160afb9c57427deb Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 5 Dec 2023 08:20:27 +0100
Subject: [PATCH 0145/1695] drivers: clk: stm32mp15: apply full clock tree in
 OP-TEE

OPTEE is now able to re-configure all clock tree.
RCC bindings is also aligned with MP13:
- merge of 'st,clksrc' and 'st,pkcs' nodes into 'st,clksrc'
- no ordering requirements on 'st,clksrc' node
- use DIV() macro for 'st,clkdiv' node
- no ordering requirements on 'st,clkdiv' node
- new pll binding

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 4e0044b2f9571363e135df63f213a7719c6e111b drivers: clk: stm32mp15: apply full clock tree in OP-TEE]
Change-Id: Ifdc18c8ece431f7a6d9a7d3126d3c85677871527
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/conf.mk |    1 +
 core/drivers/clk/clk-stm32-core.c   |    2 +
 core/drivers/clk/clk-stm32mp15.c    | 1706 ++++++++++++++++++++++++---
 3 files changed, 1526 insertions(+), 183 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/conf.mk b/core/arch/arm/plat-stm32mp1/conf.mk
index e017881a8..bac06c67a 100644
--- a/core/arch/arm/plat-stm32mp1/conf.mk
+++ b/core/arch/arm/plat-stm32mp1/conf.mk
@@ -148,6 +148,7 @@ $(call force,CFG_SECONDARY_INIT_CNTFRQ,y)
 $(call force,CFG_STM32MP1_SHARED_RESOURCES,y)
 $(call force,CFG_STM32_SAES,n)
 $(call force,CFG_STM32MP15_CLK,y)
+$(call force,CFG_STM32MP_CLK_CORE,y)
 CFG_CORE_RESERVED_SHM ?= n
 CFG_HALT_CORES_ON_PANIC ?= y
 CFG_STM32_BSEC_SIP ?= y
diff --git a/core/drivers/clk/clk-stm32-core.c b/core/drivers/clk/clk-stm32-core.c
index 90aa61f0c..4492af12b 100644
--- a/core/drivers/clk/clk-stm32-core.c
+++ b/core/drivers/clk/clk-stm32-core.c
@@ -536,6 +536,7 @@ const struct clk_ops clk_fixed_clk_ops = {
 	.get_rate	= clk_fixed_get_rate,
 };
 
+#ifndef CFG_STM32MP15
 struct clk *stm32mp_rcc_clock_id_to_clk(unsigned long clock_id)
 {
 	struct clk_stm32_priv *priv = clk_stm32_get_priv();
@@ -545,6 +546,7 @@ struct clk *stm32mp_rcc_clock_id_to_clk(unsigned long clock_id)
 
 	return priv->clk_refs[clock_id];
 }
+#endif
 
 static TEE_Result stm32mp_clk_dt_get_clk(struct dt_pargs *pargs,
 					 void *data __unused,
diff --git a/core/drivers/clk/clk-stm32mp15.c b/core/drivers/clk/clk-stm32mp15.c
index d99408c7d..7c7cafc83 100644
--- a/core/drivers/clk/clk-stm32mp15.c
+++ b/core/drivers/clk/clk-stm32mp15.c
@@ -29,9 +29,32 @@
 #include <trace.h>
 #include <util.h>
 
+#include "clk-stm32-core.h"
+
+#define MAX_OPP		(2U)
+
+#define DT_OPP_COMPAT		"operating-points-v2"
+
+/* PLL settings computation related definitions */
+#define POST_DIVM_MIN	8000000
+#define POST_DIVM_MAX	16000000
+#define DIVM_MIN	0
+#define DIVM_MAX	63
+#define DIVN_MIN	24
+#define DIVN_MAX	99
+#define DIVP_MIN	0
+#define DIVP_MAX	127
+#define FRAC_MAX	8192
+#define VCO_MIN		800000000
+#define VCO_MAX		1600000000
+
+#define PLL1_SETTINGS_VALID_ID		0x504C4C31 /* "PLL1" */
+
 static TEE_Result stm32_clock_pm(enum pm_op op, unsigned int pm_hint __unused,
 				 const struct pm_callback_handle *hdl __unused);
 
+static unsigned long clk_stm32_get_rate_oscillateur(int osc_id);
+
 /* Identifiers for root oscillators */
 enum stm32mp_osc_id {
 	OSC_HSI,
@@ -194,6 +217,16 @@ enum stm32mp1_div_id {
 	_DIV_NB,
 };
 
+enum stm32mp1_pllcfg {
+	PLLCFG_M,
+	PLLCFG_N,
+	PLLCFG_P,
+	PLLCFG_Q,
+	PLLCFG_R,
+	PLLCFG_O,
+	PLLCFG_NB
+};
+
 enum stm32mp1_plltype {
 	PLL_800,
 	PLL_1600,
@@ -246,6 +279,20 @@ struct stm32mp1_clk_pll {
 	enum stm32mp_osc_id refclk[REFCLK_SIZE];
 };
 
+struct stm32mp1_pll {
+	uint8_t refclk_min;
+	uint8_t refclk_max;
+};
+
+/* Compact structure of 32bit cells, copied raw when suspending */
+struct stm32mp1_pll_settings {
+	uint32_t valid_id;
+	uint32_t freq[MAX_OPP];
+	uint32_t volt[MAX_OPP];
+	uint32_t cfg[MAX_OPP][PLLCFG_NB];
+	uint32_t frac[MAX_OPP];
+};
+
 #define N_S	0	/* Non-secure can access RCC interface */
 #define SEC	1	/* RCC[TZEN] protects RCC interface */
 
@@ -506,6 +553,18 @@ static const struct stm32mp1_clk_sel stm32mp1_clk_sel[_PARENT_SEL_NB] = {
 	_CLK_PARENT(_MCUSS_SEL, RCC_MSSCKSELR, 0, 0x3, mcuss_parents),
 };
 
+/* Define characteristics of PLL according type */
+static const struct stm32mp1_pll stm32mp1_pll[PLL_TYPE_NB] = {
+	[PLL_800] = {
+		.refclk_min = 4,
+		.refclk_max = 16,
+	},
+	[PLL_1600] = {
+		.refclk_min = 8,
+		.refclk_max = 16,
+	},
+};
+
 /* PLLNCFGR2 register divider by output */
 static const uint8_t pllncfgr2[_DIV_NB] = {
 	[_DIV_P] = RCC_PLLNCFGR2_DIVP_SHIFT,
@@ -589,20 +648,14 @@ static const char __maybe_unused *const stm32mp1_clk_parent_name[_PARENT_NB] = {
 	[_USB_PHY_48] = "USB_PHY_48",
 };
 
-/*
- * Oscillator frequency in Hz. This array shall be initialized
- * according to platform.
- */
-static unsigned long stm32mp1_osc[NB_OSC];
-
 static unsigned long osc_frequency(enum stm32mp_osc_id idx)
 {
-	if (idx >= ARRAY_SIZE(stm32mp1_osc)) {
+	if (idx >= NB_OSC) {
 		DMSG("clk id %d not found", idx);
 		return 0;
 	}
 
-	return stm32mp1_osc[idx];
+	return clk_stm32_get_rate_oscillateur(idx);
 }
 
 static const struct stm32mp1_clk_gate *gate_ref(unsigned int idx)
@@ -1217,84 +1270,6 @@ void stm32mp_register_clock_parents_secure(unsigned long clock_id)
 	secure_parent_clocks(parent_id);
 }
 
-static const char *stm32mp_osc_node_label[NB_OSC] = {
-	[OSC_LSI] = "clk-lsi",
-	[OSC_LSE] = "clk-lse",
-	[OSC_HSI] = "clk-hsi",
-	[OSC_HSE] = "clk-hse",
-	[OSC_CSI] = "clk-csi",
-	[OSC_I2S_CKIN] = "i2s_ckin",
-	[OSC_USB_PHY_48] = "ck_usbo_48m"
-};
-
-static unsigned int clk_freq_prop(const void *fdt, int node)
-{
-	const fdt32_t *cuint = NULL;
-	int ret = 0;
-
-	/* Disabled clocks report null rate */
-	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
-		return 0;
-
-	cuint = fdt_getprop(fdt, node, "clock-frequency", &ret);
-	if (!cuint)
-		panic();
-
-	return fdt32_to_cpu(*cuint);
-}
-
-static void get_osc_freq_from_dt(const void *fdt)
-{
-	enum stm32mp_osc_id idx = _UNKNOWN_OSC_ID;
-	int clk_node = fdt_path_offset(fdt, "/clocks");
-
-	if (clk_node < 0)
-		panic();
-
-	COMPILE_TIME_ASSERT((int)OSC_HSI == 0);
-	for (idx = OSC_HSI; idx < NB_OSC; idx++) {
-		const char *name = stm32mp_osc_node_label[idx];
-		int subnode = 0;
-
-		fdt_for_each_subnode(subnode, fdt, clk_node) {
-			const char *cchar = NULL;
-			int ret = 0;
-
-			cchar = fdt_get_name(fdt, subnode, &ret);
-			if (!cchar)
-				panic();
-
-			if (strncmp(cchar, name, (size_t)ret) == 0) {
-				stm32mp1_osc[idx] = clk_freq_prop(fdt, subnode);
-
-				DMSG("Osc %s: %lu Hz", name, stm32mp1_osc[idx]);
-				break;
-			}
-		}
-
-		if (!stm32mp1_osc[idx])
-			DMSG("Osc %s: no frequency info", name);
-	}
-}
-
-static void enable_static_secure_clocks(void)
-{
-	unsigned int idx = 0;
-	const unsigned long secure_enable[] = {
-		DDRC1, DDRC1LP, DDRC2, DDRC2LP, DDRPHYC, DDRPHYCLP, DDRCAPB,
-		AXIDCG, DDRPHYCAPB, DDRPHYCAPBLP, TZPC, TZC1, TZC2, STGEN_K,
-		BSEC,
-	};
-
-	for (idx = 0; idx < ARRAY_SIZE(secure_enable); idx++) {
-		clk_enable(stm32mp_rcc_clock_id_to_clk(secure_enable[idx]));
-		stm32mp_register_clock_parents_secure(secure_enable[idx]);
-	}
-
-	if (CFG_TEE_CORE_NB_CORE > 1)
-		clk_enable(stm32mp_rcc_clock_id_to_clk(RTCAPB));
-}
-
 static void __maybe_unused enable_rcc_tzen(void)
 {
 	io_setbits32(stm32_rcc_base() + RCC_TZCR, RCC_TZCR_TZEN);
@@ -1306,55 +1281,6 @@ static void __maybe_unused disable_rcc_tzen(void)
 	io_clrbits32(stm32_rcc_base() + RCC_TZCR, RCC_TZCR_TZEN);
 }
 
-static TEE_Result stm32mp1_clk_fdt_init(const void *fdt, int node)
-{
-	unsigned int i = 0;
-	int len = 0;
-	int ignored = 0;
-
-	get_osc_freq_from_dt(fdt);
-
-	/*
-	 * OP-TEE core is not in charge of configuring clock parenthood.
-	 * This is expected from an earlier boot stage. Modifying the clock
-	 * tree parenthood here may jeopardize already configured clocks.
-	 * The sequence below ignores such DT directives with a friendly
-	 * debug trace.
-	 */
-	if (fdt_getprop(fdt, node, "st,clksrc", &len)) {
-		DMSG("Ignore source clocks configuration from DT");
-		ignored++;
-	}
-	if (fdt_getprop(fdt, node, "st,clkdiv", &len)) {
-		DMSG("Ignore clock divisors configuration from DT");
-		ignored++;
-	}
-	if (fdt_getprop(fdt, node, "st,pkcs", &len)) {
-		DMSG("Ignore peripheral clocks tree configuration from DT");
-		ignored++;
-	}
-	for (i = (enum stm32mp1_pll_id)0; i < _PLL_NB; i++) {
-		char name[] = "st,pll@X";
-		int subnode = -1;
-
-		snprintf(name, sizeof(name), "st,pll@%d", i);
-		subnode = fdt_subnode_offset(fdt, node, name);
-		if (subnode < 0)
-			continue;
-
-		if (fdt_getprop(fdt, subnode, "cfg", &len) ||
-		    fdt_getprop(fdt, subnode, "frac", &len)) {
-			DMSG("Ignore PLL%u configurations from DT", i);
-			ignored++;
-		}
-	}
-
-	if (ignored != 0)
-		IMSG("DT clock tree configurations were ignored");
-
-	return TEE_SUCCESS;
-}
-
 void stm32mp1_clk_mcuss_protect(bool enable)
 {
 	uintptr_t rcc_base = stm32_rcc_base();
@@ -1592,79 +1518,1493 @@ static TEE_Result stm32mp1_clk_dt_get_clk(struct dt_pargs *pargs,
 /* Non-null reference for compat data */
 static const uint8_t non_secure_rcc;
 
-static TEE_Result stm32mp1_clock_provider_probe(const void *fdt, int offs,
-						const void *compat_data)
+static void enable_static_secure_clocks(void)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int idx = 0;
+	const unsigned long secure_enable[] = {
+		DDRC1, DDRC1LP, DDRC2, DDRC2LP, DDRPHYC, DDRPHYCLP, DDRCAPB,
+		AXIDCG, DDRPHYCAPB, DDRPHYCAPBLP, TZPC, TZC1, TZC2, STGEN_K,
+		BSEC,
+	};
 
-	if (compat_data == &non_secure_rcc)
-		disable_rcc_tzen();
-	else
-		enable_rcc_tzen();
+	for (idx = 0; idx < ARRAY_SIZE(secure_enable); idx++) {
+		clk_enable(stm32mp_rcc_clock_id_to_clk(secure_enable[idx]));
+		stm32mp_register_clock_parents_secure(secure_enable[idx]);
+	}
 
-	res = stm32mp1_clk_fdt_init(fdt, offs);
-	if (res) {
-		EMSG("Failed to initialize clocks from DT: %#"PRIx32, res);
-		panic();
+	if (CFG_TEE_CORE_NB_CORE > 1)
+		clk_enable(stm32mp_rcc_clock_id_to_clk(RTCAPB));
+}
+
+struct stm32_osci_dt_cfg {
+	unsigned long freq;
+	bool bypass;
+	bool digbyp;
+	bool css;
+	uint32_t drive;
+};
+
+enum pll_mn {
+	PLL_CFG_M,
+	PLL_CFG_N,
+	PLL_DIV_MN_NB
+};
+
+enum pll_pqr {
+	PLL_CFG_P,
+	PLL_CFG_Q,
+	PLL_CFG_R,
+	PLL_DIV_PQR_NB
+};
+
+enum pll_csg {
+	PLL_CSG_MOD_PER,
+	PLL_CSG_INC_STEP,
+	PLL_CSG_SSCG_MODE,
+	PLL_CSG_NB
+};
+
+struct stm32_pll_vco {
+	uint32_t status;
+	uint32_t src;
+	uint32_t div_mn[PLL_DIV_MN_NB];
+	uint32_t frac;
+	bool csg_enabled;
+	uint32_t csg[PLL_CSG_NB];
+};
+
+struct stm32_pll_output {
+	uint32_t output[PLL_DIV_PQR_NB];
+};
+
+struct stm32_pll_dt_cfg {
+	struct stm32_pll_vco vco;
+	struct stm32_pll_output output;
+};
+
+struct stm32_clk_opp_cfg {
+	uint32_t frq;
+	uint32_t src;
+	uint32_t div;
+	struct stm32_pll_dt_cfg pll_cfg;
+};
+
+struct stm32_clk_opp_dt_cfg {
+	struct stm32_clk_opp_cfg mpu_opp[MAX_OPP];
+	struct stm32_clk_opp_cfg axi_opp[MAX_OPP];
+	struct stm32_clk_opp_cfg mcu_opp[MAX_OPP];
+};
+
+struct stm32_clk_platdata {
+	uintptr_t rcc_base;
+	uint32_t nosci;
+	struct stm32_osci_dt_cfg *osci;
+	uint32_t npll;
+	struct stm32_pll_dt_cfg *pll;
+	struct stm32_clk_opp_dt_cfg *opp;
+	uint32_t nclksrc;
+	uint32_t *clksrc;
+	uint32_t nclkdiv;
+	uint32_t *clkdiv;
+	struct stm32_pinctrl_list *pinctrl_cfg;
+};
+
+/*
+ * GATE CONFIG
+ */
+
+/* Warning GATE_XXX_RDY must follow GATE_XXX */
+enum enum_gate_cfg {
+	GATE_LSE,
+	GATE_LSE_RDY,
+	GATE_LSI,
+	GATE_LSI_RDY,
+	GATE_HSI,
+	GATE_HSI_RDY,
+	GATE_CSI,
+	GATE_CSI_RDY,
+	GATE_HSE,
+	GATE_HSE_RDY,
+	GATE_PLL1,
+	GATE_PLL1_RDY,
+	GATE_PLL2,
+	GATE_PLL2_RDY,
+	GATE_PLL3,
+	GATE_PLL3_RDY,
+	GATE_PLL4,
+	GATE_PLL4_RDY,
+	GATE_HSIDIVRDY,
+	GATE_MPUSRCRDY,
+	GATE_AXISSRCRDY,
+	GATE_MCUSSRCRDY,
+	GATE_PLL12SRCRDY,
+	GATE_PLL3SRCRDY,
+	GATE_PLL4SRCRDY,
+	GATE_MPUDIVRDY,
+	GATE_AXIDIVRDY,
+	GATE_MCUDIVRDY,
+	GATE_APB1DIVRDY,
+	GATE_APB2DIVRDY,
+	GATE_APB3DIVRDY,
+	GATE_APB4DIVRDY,
+	GATE_APB5DIVRDY,
+	GATE_RTCCK,
+	GATE_MCO1,
+	GATE_MCO2,
+	GATE_NB
+};
+
+#define GATE_CFG(_id, _offset, _bit_idx, _offset_clr)\
+	[(_id)] = {\
+		.offset		= (_offset),\
+		.bit_idx	= (_bit_idx),\
+		.set_clr	= (_offset_clr),\
 	}
 
-	res = register_stm32mp1_clocks();
-	if (res) {
-		EMSG("Failed to register clocks: %#"PRIx32, res);
-		panic();
+static const struct gate_cfg gates_mp15[GATE_NB] = {
+	GATE_CFG(GATE_LSE,		RCC_BDCR,		0,	0),
+	GATE_CFG(GATE_LSE_RDY,		RCC_BDCR,		2,	0),
+	GATE_CFG(GATE_RTCCK,		RCC_BDCR,		20,	0),
+	GATE_CFG(GATE_LSI,		RCC_RDLSICR,		0,	0),
+	GATE_CFG(GATE_LSI_RDY,		RCC_RDLSICR,		1,	0),
+	GATE_CFG(GATE_HSI,		RCC_OCENSETR,		0,	1),
+	GATE_CFG(GATE_HSI_RDY,		RCC_OCRDYR,		0,	0),
+	GATE_CFG(GATE_CSI,		RCC_OCENSETR,		4,	1),
+	GATE_CFG(GATE_CSI_RDY,		RCC_OCRDYR,		4,	0),
+	GATE_CFG(GATE_HSE,		RCC_OCENSETR,		8,	1),
+	GATE_CFG(GATE_HSE_RDY,		RCC_OCRDYR,		8,	0),
+	GATE_CFG(GATE_HSIDIVRDY,	RCC_OCRDYR,		2,	0),
+	GATE_CFG(GATE_MPUSRCRDY,	RCC_MPCKSELR,		31,	0),
+	GATE_CFG(GATE_AXISSRCRDY,	RCC_ASSCKSELR,		31,	0),
+	GATE_CFG(GATE_MCUSSRCRDY,	RCC_MSSCKSELR,		31,	0),
+	GATE_CFG(GATE_PLL12SRCRDY,	RCC_RCK12SELR,		31,	0),
+	GATE_CFG(GATE_PLL3SRCRDY,	RCC_RCK3SELR,		31,	0),
+	GATE_CFG(GATE_PLL4SRCRDY,	RCC_RCK4SELR,		31,	0),
+	GATE_CFG(GATE_MPUDIVRDY,	RCC_MPCKDIVR,		31,	0),
+	GATE_CFG(GATE_AXIDIVRDY,	RCC_AXIDIVR,		31,	0),
+	GATE_CFG(GATE_MCUDIVRDY,	RCC_MCUDIVR,		31,	0),
+	GATE_CFG(GATE_APB1DIVRDY,	RCC_APB1DIVR,		31,	0),
+	GATE_CFG(GATE_APB2DIVRDY,	RCC_APB2DIVR,		31,	0),
+	GATE_CFG(GATE_APB3DIVRDY,	RCC_APB3DIVR,		31,	0),
+	GATE_CFG(GATE_APB4DIVRDY,	RCC_APB4DIVR,		31,	0),
+	GATE_CFG(GATE_APB5DIVRDY,	RCC_APB5DIVR,		31,	0),
+	GATE_CFG(GATE_PLL1,		RCC_PLL1CR,		0,	0),
+	GATE_CFG(GATE_PLL1_RDY,		RCC_PLL1CR,		1,	0),
+	GATE_CFG(GATE_PLL2,		RCC_PLL2CR,		0,	0),
+	GATE_CFG(GATE_PLL2_RDY,		RCC_PLL2CR,		1,	0),
+	GATE_CFG(GATE_PLL3,		RCC_PLL3CR,		0,	0),
+	GATE_CFG(GATE_PLL3_RDY,		RCC_PLL3CR,		1,	0),
+	GATE_CFG(GATE_PLL4,		RCC_PLL4CR,		0,	0),
+	GATE_CFG(GATE_PLL4_RDY,		RCC_PLL4CR,		1,	0),
+	GATE_CFG(GATE_MCO1,		RCC_MCO1CFGR,		12,	0),
+	GATE_CFG(GATE_MCO2,		RCC_MCO2CFGR,		12,	0),
+};
+
+/*
+ * DIV CONFIG
+ */
+
+static const struct div_table_cfg axi_div_table[] = {
+	{ 0, 1 }, { 1, 2 }, { 2, 3 }, { 3, 4 },
+	{ 4, 4 }, { 5, 4 }, { 6, 4 }, { 7, 4 },
+	{ 0 },
+};
+
+static const struct div_table_cfg mcu_div_table[] = {
+	{ 0, 1 }, { 1, 2 }, { 2, 4 }, { 3, 8 },
+	{ 4, 16 }, { 5, 32 }, { 6, 64 }, { 7, 128 },
+	{ 8, 256 }, { 9, 512 }, { 10, 512}, { 11, 512 },
+	{ 12, 512 }, { 13, 512 }, { 14, 512}, { 15, 512 },
+	{ 0 },
+};
+
+static const struct div_table_cfg apb_div_table[] = {
+	{ 0, 1 }, { 1, 2 }, { 2, 4 }, { 3, 8 },
+	{ 4, 16 }, { 5, 16 }, { 6, 16 }, { 7, 16 },
+	{ 0 },
+};
+
+static const struct div_table_cfg ck_trace_div_table[] = {
+	{ 0, 1 }, { 1, 2 }, { 2, 4 }, { 3, 8 },
+	{ 4, 16 }, { 5, 16 }, { 6, 16 }, { 7, 16 },
+	{ 0 },
+};
+
+#define DIVRDY_CFG(_id, _offset, _shift, _width, _flags, _table, _ready)\
+	[(_id)] = {\
+		.offset	= (_offset),\
+		.shift	= (_shift),\
+		.width	= (_width),\
+		.flags	= (_flags),\
+		.table	= (_table),\
+		.ready	= (_ready),\
 	}
 
-	res = clk_dt_register_clk_provider(fdt, offs, stm32mp1_clk_dt_get_clk,
-					   NULL);
-	if (res) {
-		EMSG("Failed to register clock provider: %#"PRIx32, res);
-		panic();
+#define DIV_CFG(_id, _offset, _shift, _width, _flags, _table)\
+	DIVRDY_CFG(_id, _offset, _shift, _width, _flags, _table, DIV_NO_RDY)
+
+static const struct div_cfg dividers_mp15[] = {
+	DIVRDY_CFG(DIV_MPU, RCC_MPCKDIVR, 0, 4, 0, NULL,
+		   GATE_MPUDIVRDY),
+	DIVRDY_CFG(DIV_AXI, RCC_AXIDIVR, 0, 3, 0, axi_div_table,
+		   GATE_AXIDIVRDY),
+	DIVRDY_CFG(DIV_MCU,	RCC_MCUDIVR, 0, 4, 0, mcu_div_table,
+		   GATE_MCUDIVRDY),
+	DIVRDY_CFG(DIV_APB1, RCC_APB1DIVR, 0, 3, 0, apb_div_table,
+		   GATE_APB1DIVRDY),
+	DIVRDY_CFG(DIV_APB2, RCC_APB2DIVR, 0, 3, 0, apb_div_table,
+		   GATE_APB2DIVRDY),
+	DIVRDY_CFG(DIV_APB3, RCC_APB3DIVR, 0, 3, 0, apb_div_table,
+		   GATE_APB3DIVRDY),
+	DIVRDY_CFG(DIV_APB4, RCC_APB4DIVR, 0, 3, 0, apb_div_table,
+		   GATE_APB4DIVRDY),
+	DIVRDY_CFG(DIV_APB5, RCC_APB5DIVR, 0, 3, 0, apb_div_table,
+		   GATE_APB5DIVRDY),
+	DIVRDY_CFG(DIV_HSI,	RCC_HSICFGR, 0, 2, CLK_DIVIDER_POWER_OF_TWO,
+		   NULL, GATE_HSIDIVRDY),
+	DIV_CFG(DIV_RTC,	RCC_RTCDIVR, 0, 6, 0, NULL),
+	DIV_CFG(DIV_MCO1,	RCC_MCO1CFGR, 4, 4, 0, NULL),
+	DIV_CFG(DIV_MCO2,	RCC_MCO2CFGR, 4, 4, 0, NULL),
+	DIV_CFG(DIV_TRACE,	RCC_DBGCFGR, 0, 3, 0, ck_trace_div_table),
+	DIV_CFG(DIV_ETHPTP,	RCC_ETHCKSELR, 4, 4, 0, NULL),
+};
+
+/*
+ * MUX CONFIG
+ */
+#define MUXRDY_CFG(_id, _offset, _shift, _witdh, _rdy)\
+	[(_id)] = {\
+			.offset	= (_offset),\
+			.shift	= (_shift),\
+			.width	= (_witdh),\
+			.ready	= (_rdy),\
 	}
 
-	enable_static_secure_clocks();
+#define MUX_CFG(_id, _offset, _shift, _witdh)\
+	MUXRDY_CFG(_id, _offset, _shift, _witdh, MUX_NO_RDY)
+
+static const struct mux_cfg parent_mp15[MUX_NB] = {
+	MUX_CFG(MUX_PLL12,	RCC_RCK12SELR, 0, 2),
+	MUX_CFG(MUX_PLL3,	RCC_RCK3SELR, 0, 2),
+	MUX_CFG(MUX_PLL4,	RCC_RCK4SELR, 0, 2),
+	MUX_CFG(MUX_CKPER,	RCC_CPERCKSELR, 0, 2),
+	MUXRDY_CFG(MUX_MPU,	RCC_MPCKSELR, 0, 2, GATE_MPUSRCRDY),
+	MUXRDY_CFG(MUX_AXI,	RCC_ASSCKSELR, 0, 3, GATE_AXISSRCRDY),
+	MUXRDY_CFG(MUX_MCU,	RCC_MSSCKSELR, 0, 2, GATE_MCUSSRCRDY),
+	MUX_CFG(MUX_RTC,	RCC_BDCR, 16, 2),
+	MUX_CFG(MUX_SDMMC12,	RCC_SDMMC12CKSELR, 0, 3),
+	MUX_CFG(MUX_SPI2S23,	RCC_SPI2S23CKSELR, 0, 3),
+	MUX_CFG(MUX_SPI45,	RCC_SPI45CKSELR, 0, 3),
+	MUX_CFG(MUX_I2C12,	RCC_I2C12CKSELR, 0, 3),
+	MUX_CFG(MUX_I2C35,	RCC_I2C35CKSELR, 0, 3),
+	MUX_CFG(MUX_LPTIM23,	RCC_LPTIM23CKSELR, 0, 3),
+	MUX_CFG(MUX_LPTIM45,	RCC_LPTIM45CKSELR, 0, 3),
+	MUX_CFG(MUX_UART24,	RCC_UART24CKSELR, 0, 3),
+	MUX_CFG(MUX_UART35,	RCC_UART35CKSELR, 0, 3),
+	MUX_CFG(MUX_UART78,	RCC_UART78CKSELR, 0, 3),
+	MUX_CFG(MUX_SAI1,	RCC_SAI1CKSELR, 0, 3),
+	MUX_CFG(MUX_ETH,	RCC_ETHCKSELR, 0, 2),
+	MUX_CFG(MUX_I2C46,	RCC_I2C46CKSELR, 0, 3),
+	MUX_CFG(MUX_RNG2,	RCC_RNG2CKSELR, 0, 2),
+	MUX_CFG(MUX_SDMMC3,	RCC_SDMMC3CKSELR, 0, 3),
+	MUX_CFG(MUX_FMC,	RCC_FMCCKSELR, 0, 2),
+	MUX_CFG(MUX_QSPI,	RCC_QSPICKSELR, 0, 2),
+	MUX_CFG(MUX_USBPHY,	RCC_USBCKSELR, 0, 2),
+	MUX_CFG(MUX_USBO,	RCC_USBCKSELR, 4, 1),
+	MUX_CFG(MUX_SPDIF,	RCC_SPDIFCKSELR, 0, 2),
+	MUX_CFG(MUX_SPI2S1,	RCC_SPI2S1CKSELR, 0, 3),
+	MUX_CFG(MUX_CEC,	RCC_CECCKSELR, 0, 2),
+	MUX_CFG(MUX_LPTIM1,	RCC_LPTIM1CKSELR, 0, 3),
+	MUX_CFG(MUX_UART6,	RCC_UART6CKSELR, 0, 3),
+	MUX_CFG(MUX_FDCAN,	RCC_FDCANCKSELR, 0, 2),
+	MUX_CFG(MUX_SAI2,	RCC_SAI2CKSELR, 0, 3),
+	MUX_CFG(MUX_SAI3,	RCC_SAI3CKSELR, 0, 3),
+	MUX_CFG(MUX_SAI4,	RCC_SAI4CKSELR, 0, 3),
+	MUX_CFG(MUX_ADC,	RCC_ADCCKSELR, 0, 2),
+	MUX_CFG(MUX_DSI,	RCC_DSICKSELR, 0, 1),
+	MUX_CFG(MUX_RNG1,	RCC_RNG1CKSELR, 0, 2),
+	MUX_CFG(MUX_STGEN,	RCC_STGENCKSELR, 0, 2),
+	MUX_CFG(MUX_UART1,	RCC_UART1CKSELR, 0, 3),
+	MUX_CFG(MUX_SPI6,	RCC_SPI6CKSELR, 0, 3),
+	MUX_CFG(MUX_MCO1,	RCC_MCO1CFGR, 0, 3),
+	MUX_CFG(MUX_MCO2,	RCC_MCO2CFGR, 0, 3),
+};
 
-	register_pm_core_service_cb(stm32_clock_pm, NULL,
-				    "stm32mp15-clk-service");
+struct stm32_osc_cfg {
+	int osc_id;
+};
 
-	return TEE_SUCCESS;
-}
+struct clk_stm32_bypass {
+	uint16_t offset;
+	uint8_t bit_byp;
+	uint8_t bit_digbyp;
+};
 
-static const struct dt_device_match stm32mp1_clock_match_table[] = {
-	{  .compatible = "st,stm32mp1-rcc", .compat_data = &non_secure_rcc, },
-	{  .compatible = "st,stm32mp1-rcc-secure", },
-	{ }
+struct clk_stm32_css {
+	uint16_t offset;
+	uint8_t bit_css;
 };
 
-DEFINE_DT_DRIVER(stm32mp1_clock_dt_driver) = {
-	.name = "stm32mp1_clock",
-	.type = DT_DRIVER_CLK,
-	.match_table = stm32mp1_clock_match_table,
-	.probe = stm32mp1_clock_provider_probe,
+struct clk_stm32_drive {
+	uint16_t offset;
+	uint8_t drv_shift;
+	uint8_t drv_width;
+	uint8_t drv_default;
+};
+
+struct clk_oscillator_data {
+	const char *name;
+	uint16_t gate_id;
+	struct clk_stm32_bypass *bypass;
+	struct clk_stm32_css *css;
+	struct clk_stm32_drive *drive;
 };
 
-#define CLKDIV_TIMEOUT_US	(200 * 1000)
+#define BYPASS(_offset, _bit_byp, _bit_digbyp) (&(struct clk_stm32_bypass){\
+	.offset		= (_offset),\
+	.bit_byp	= (_bit_byp),\
+	.bit_digbyp	= (_bit_digbyp),\
+})
+
+#define CSS(_offset, _bit_css) (&(struct clk_stm32_css){\
+	.offset		= (_offset),\
+	.bit_css	= (_bit_css),\
+})
+
+#define DRIVE(_offset, _shift, _width, _default) (&(struct clk_stm32_drive){\
+	.offset		= (_offset),\
+	.drv_shift	= (_shift),\
+	.drv_width	= (_width),\
+	.drv_default	= (_default),\
+})
+
+#define OSCILLATOR(idx_osc, _name, _gate_id, _bypass, _css, _drive) \
+	[(idx_osc)] = (struct clk_oscillator_data){\
+		.name		= (_name),\
+		.gate_id	= (_gate_id),\
+		.bypass		= (_bypass),\
+		.css		= (_css),\
+		.drive		= (_drive),\
+	}
+
+static const struct clk_oscillator_data stm32mp15_osc_data[NB_OSC] = {
+	OSCILLATOR(OSC_HSI, "clk-hsi", GATE_HSI,
+		   NULL, NULL, NULL),
+
+	OSCILLATOR(OSC_LSI, "clk-lsi", GATE_LSI,
+		   NULL, NULL, NULL),
 
-static int stm32mp1_set_clkdiv(unsigned int clkdiv, uintptr_t address)
+	OSCILLATOR(OSC_CSI, "clk-csi", GATE_CSI,
+		   NULL, NULL, NULL),
+
+	OSCILLATOR(OSC_LSE, "clk-lse", GATE_LSE,
+		   BYPASS(RCC_BDCR, 1, 3),
+		   CSS(RCC_BDCR, 8),
+		   DRIVE(RCC_BDCR, 4, 2, 2)),
+
+	OSCILLATOR(OSC_HSE, "clk-hse", GATE_HSE,
+		   BYPASS(RCC_OCENSETR, 10, 7),
+		   CSS(RCC_OCENSETR, 11),
+		   NULL),
+	OSCILLATOR(OSC_I2S_CKIN, "i2s_ckin", -1, NULL, NULL, NULL),
+	OSCILLATOR(OSC_USB_PHY_48, "ck_usbo_48m", -1, NULL, NULL, NULL),
+};
+
+static const struct clk_oscillator_data *clk_oscillator_get_data(int osc_id)
 {
-	uint64_t timeout_ref = 0;
+	assert(osc_id >= 0 && osc_id < (int)ARRAY_SIZE(stm32mp15_osc_data));
 
-	io_clrsetbits32(address, RCC_DIVR_DIV_MASK, clkdiv & RCC_DIVR_DIV_MASK);
+	return &stm32mp15_osc_data[osc_id];
+}
 
-	timeout_ref = timeout_init_us(CLKDIV_TIMEOUT_US);
-	while (!(io_read32(address) & RCC_DIVR_DIVRDY))
-		if (timeout_elapsed(timeout_ref))
-			break;
+static unsigned long clk_stm32_get_rate_oscillateur(int osc_id)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[osc_id];
+
+	return osci->freq;
+}
 
-	if (!(io_read32(address) & RCC_DIVR_DIVRDY)) {
-		EMSG("CLKDIV 0x%x start failed @ 0x%"PRIxPTR": 0x%"PRIx32,
-		     clkdiv, address, io_read32(address));
+static void
+clk_oscillator_set_bypass(struct clk_stm32_priv *priv,
+			  const struct clk_oscillator_data *osc_data,
+			  bool digbyp, bool bypass)
+{
+	const struct clk_stm32_bypass *bypass_data = osc_data->bypass;
+	uintptr_t address = 0;
 
-		return -1;
+	if (!bypass_data)
+		return;
+
+	address = priv->base + bypass_data->offset;
+
+	if (digbyp)
+		io_setbits32(address, BIT(bypass_data->bit_digbyp));
+
+	if (bypass || digbyp)
+		io_setbits32(address, BIT(bypass_data->bit_byp));
+}
+
+static void clk_oscillator_set_css(struct clk_stm32_priv *priv,
+				   const struct clk_oscillator_data *osc_data,
+				   bool css)
+{
+	struct clk_stm32_css *css_data = osc_data->css;
+	uintptr_t address = 0;
+
+	if (!css_data)
+		return;
+
+	address = priv->base + css_data->offset;
+
+	if (css)
+		io_setbits32(address, BIT(css_data->bit_css));
+}
+
+static void clk_oscillator_set_drive(struct clk_stm32_priv *priv,
+				     const struct clk_oscillator_data *osc_data,
+				     uint8_t lsedrv)
+{
+	struct clk_stm32_drive *drive_data = osc_data->drive;
+	uintptr_t address = 0;
+	uint32_t mask = 0;
+	uint32_t value = 0;
+
+	if (!drive_data)
+		return;
+
+	address = priv->base + drive_data->offset;
+
+	mask = (BIT(drive_data->drv_width) - 1U) << drive_data->drv_shift;
+
+	/*
+	 * Warning: not recommended to switch directly from "high drive"
+	 * to "medium low drive", and vice-versa.
+	 */
+	value = (io_read32(address) & mask) >> drive_data->drv_shift;
+
+	while (value != lsedrv) {
+		if (value > lsedrv)
+			value--;
+		else
+			value++;
+
+		io_clrsetbits32(address, mask, value << drive_data->drv_shift);
 	}
+}
 
-	return 0;
+static void stm32_enable_oscillator_hse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_HSE];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	if (osci->freq == 0U)
+		return;
+
+	osc_data = clk_oscillator_get_data(OSC_HSE);
+
+	clk_oscillator_set_bypass(priv, osc_data,  osci->digbyp, osci->bypass);
+
+	/* Enable clock and wait ready bit */
+	if (stm32_gate_rdy_enable(osc_data->gate_id)) {
+		EMSG("timeout to enable hse clock");
+		panic();
+	}
+
+	clk_oscillator_set_css(priv, osc_data, osci->css);
+}
+
+static void stm32_enable_oscillator_lse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	if (osci->freq == 0U)
+		return;
+
+	osc_data = clk_oscillator_get_data(OSC_LSE);
+
+	if (stm32_gate_is_enabled(osc_data->gate_id))
+		return;
+
+	clk_oscillator_set_bypass(priv, osc_data, osci->digbyp, osci->bypass);
+
+	clk_oscillator_set_drive(priv, osc_data,  osci->drive);
+
+	/* Enable lse clock, but don't wait ready bit */
+	stm32_gate_enable(osc_data->gate_id);
+}
+
+static void
+stm32_enable_oscillator_lsi(struct clk_stm32_priv *priv __maybe_unused,
+			    struct stm32_clk_platdata *pdata)
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSI];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	if (osci->freq == 0U)
+		return;
+
+	osc_data = clk_oscillator_get_data(OSC_LSI);
+
+	/* Enable clock and wait ready bit */
+	if (stm32_gate_rdy_enable(osc_data->gate_id)) {
+		EMSG("timeout to enable lsi clock");
+		panic();
+	}
 }
 
+static void
+stm32_enable_oscillator_csi(struct clk_stm32_priv *priv __maybe_unused,
+			    struct stm32_clk_platdata *pdata)
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_CSI];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	if (osci->freq == 0U)
+		return;
+
+	osc_data = clk_oscillator_get_data(OSC_CSI);
+
+	/* Enable clock and wait ready bit */
+	if (stm32_gate_rdy_enable(osc_data->gate_id)) {
+		EMSG("timeout to enable csi clock");
+		panic();
+	}
+}
+
+static int stm32_clk_oscillators_lse_set_css(struct clk_stm32_priv *priv,
+					     struct stm32_clk_platdata *pdata)
+
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	osc_data = clk_oscillator_get_data(OSC_LSE);
+
+	clk_oscillator_set_css(priv, osc_data, osci->css);
+
+	return 0;
+}
+
+static int
+stm32_clk_oscillators_wait_lse_ready(struct clk_stm32_priv *priv __maybe_unused,
+				     struct stm32_clk_platdata *pdata)
+{
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+	const struct clk_oscillator_data *osc_data = NULL;
+
+	if (osci->freq == 0U)
+		return 0;
+
+	osc_data = clk_oscillator_get_data(OSC_LSE);
+
+	if (stm32_gate_wait_ready(osc_data->gate_id, true))
+		return -1;
+
+	return 0;
+}
+
+static void stm32_clk_oscillators_enable(struct clk_stm32_priv *priv,
+					 struct stm32_clk_platdata *pdata)
+{
+	stm32_enable_oscillator_hse(priv, pdata);
+	stm32_enable_oscillator_lse(priv, pdata);
+	stm32_enable_oscillator_lsi(priv, pdata);
+	stm32_enable_oscillator_csi(priv, pdata);
+}
+
+#define RCC_OFFSET_PLLXCR		0
+#define RCC_OFFSET_PLLXCFGR1		4
+#define RCC_OFFSET_PLLXCFGR2		8
+#define RCC_OFFSET_PLLXFRACR		12
+#define RCC_OFFSET_PLLXCSGR		16
+
+struct stm32_clk_pll {
+	enum stm32mp1_plltype plltype;
+	uint16_t gate_id;
+	uint16_t mux_id;
+	uint16_t reg_pllxcr;
+};
+
+#define CLK_PLL_CFG(_idx, _type, _gate_id, _mux_id, _reg)\
+	[(_idx)] = {\
+		.gate_id = (_gate_id),\
+		.mux_id = (_mux_id),\
+		.plltype = (_type),\
+		.reg_pllxcr = (_reg),\
+	}
+
+static const struct stm32_clk_pll stm32_mp15_clk_pll[_PLL_NB] = {
+	CLK_PLL_CFG(_PLL1, PLL_1600, GATE_PLL1, MUX_PLL12, RCC_PLL1CR),
+	CLK_PLL_CFG(_PLL2, PLL_1600, GATE_PLL2, MUX_PLL12, RCC_PLL2CR),
+	CLK_PLL_CFG(_PLL3, PLL_800, GATE_PLL3, MUX_PLL3, RCC_PLL3CR),
+	CLK_PLL_CFG(_PLL4, PLL_800, GATE_PLL4, MUX_PLL4, RCC_PLL4CR),
+};
+
+static const struct stm32_clk_pll *clk_stm32_pll_data(unsigned int idx)
+{
+	return &stm32_mp15_clk_pll[idx];
+}
+
+static int fdt_clk_stm32_parse_oscillator(const void *fdt, int node,
+					  const char *name,
+					  struct stm32_osci_dt_cfg *osci)
+{
+	int subnode = 0;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		const char *cchar = NULL;
+		const fdt32_t *cuint = NULL;
+		int ret = 0;
+
+		cchar = fdt_get_name(fdt, subnode, &ret);
+		if (!cchar)
+			return ret;
+
+		if (strncmp(cchar, name, (size_t)ret) ||
+		    fdt_get_status(fdt, subnode) == DT_STATUS_DISABLED)
+			continue;
+
+		cuint = fdt_getprop(fdt, subnode, "clock-frequency", &ret);
+		if (!cuint)
+			panic();
+
+		osci->freq = fdt32_to_cpu(*cuint);
+
+		if (fdt_getprop(fdt, subnode, "st,bypass", NULL))
+			osci->bypass = true;
+
+		if (fdt_getprop(fdt, subnode, "st,digbypass", NULL))
+			osci->digbyp = true;
+
+		if (fdt_getprop(fdt, subnode, "st,css", NULL))
+			osci->css = true;
+
+		osci->drive = fdt_read_uint32_default(fdt, subnode, "st,drive",
+						      LSEDRV_MEDIUM_HIGH);
+
+		return 0;
+	}
+
+	return -FDT_ERR_NOTFOUND;
+}
+
+static int fdt_stm32_clk_parse_all_oscillator(const void *fdt,
+					      int node __maybe_unused,
+					      struct stm32_clk_platdata *pdata)
+{
+	int fdt_err = 0;
+	size_t i = 0;
+	int osc_node = 0;
+
+	osc_node = fdt_path_offset(fdt, "/clocks");
+	if (osc_node < 0)
+		return -FDT_ERR_NOTFOUND;
+
+	for (i = 0; i < NB_OSC; i++) {
+		struct stm32_osci_dt_cfg *osci = &pdata->osci[i];
+		const struct clk_oscillator_data *osc_data = NULL;
+
+		osc_data = clk_oscillator_get_data(i);
+
+		fdt_err = fdt_clk_stm32_parse_oscillator(fdt, osc_node,
+							 osc_data->name, osci);
+		if (fdt_err < 0)
+			osci->freq = 0UL;
+	}
+
+	return 0;
+}
+
+static int fdt_clk_stm32_load_vco_config(const void *fdt, int subnode,
+					 struct stm32_pll_vco *vco)
+{
+	int ret = 0;
+
+	ret = fdt_read_uint32_array(fdt, subnode, "divmn", vco->div_mn,
+				    PLL_DIV_MN_NB);
+	if (ret != 0)
+		return ret;
+
+	ret = fdt_read_uint32_array(fdt, subnode, "csg", vco->csg,
+				    PLL_CSG_NB);
+
+	vco->csg_enabled = (ret == 0);
+
+	if (ret == -FDT_ERR_NOTFOUND)
+		ret = 0;
+
+	if (ret != 0)
+		return ret;
+
+	vco->status = RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
+		      RCC_PLLNCR_DIVREN | RCC_PLLNCR_PLLON;
+
+	vco->frac = fdt_read_uint32_default(fdt, subnode, "frac", 0);
+
+	vco->src = fdt_read_uint32_default(fdt, subnode, "src", UINT32_MAX);
+
+	return 0;
+}
+
+static int fdt_clk_stm32_load_output_config(const void *fdt, int subnode,
+					    struct stm32_pll_output *output)
+{
+	return fdt_read_uint32_array(fdt, subnode, "st,pll_div_pqr",
+				     output->output, (int)PLL_DIV_PQR_NB);
+}
+
+static int fdt_clk_stm32_parse_pll(const void *fdt, int subnode,
+				   struct stm32_pll_dt_cfg *pll)
+{
+	const fdt32_t *cuint = NULL;
+	int subnode_pll = 0;
+	int subnode_vco = 0;
+	int err = 0;
+
+	cuint = fdt_getprop(fdt, subnode, "st,pll", NULL);
+	if (!cuint)
+		return -FDT_ERR_NOTFOUND;
+
+	subnode_pll = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(*cuint));
+	if (subnode_pll < 0)
+		return -FDT_ERR_NOTFOUND;
+
+	cuint = fdt_getprop(fdt, subnode_pll, "st,pll_vco", NULL);
+	if (!cuint)
+		return -FDT_ERR_NOTFOUND;
+
+	subnode_vco = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(*cuint));
+	if (subnode_vco < 0)
+		return -FDT_ERR_NOTFOUND;
+
+	err = fdt_clk_stm32_load_vco_config(fdt, subnode_vco, &pll->vco);
+	if (err != 0)
+		return err;
+
+	err = fdt_clk_stm32_load_output_config(fdt, subnode_pll, &pll->output);
+	if (err != 0)
+		return err;
+
+	return 0;
+}
+
+#define RCC_PLL_NAME_SIZE	12
+
+static int fdt_stm32_clk_parse_all_pll(const void *fdt, int node,
+				       struct stm32_clk_platdata *pdata)
+{
+	size_t i = 0;
+
+	for (i = _PLL1; i < pdata->npll; i++) {
+		struct stm32_pll_dt_cfg *pll = pdata->pll + i;
+		char name[RCC_PLL_NAME_SIZE] = { 0 };
+		int subnode = 0;
+		int err = 0;
+
+		err = snprintf(name, sizeof(name), "st,pll@%d", i);
+		if (err < 0 || (size_t)err >= sizeof(name))
+			panic();
+
+		subnode = fdt_subnode_offset(fdt, node, name);
+		if (subnode < 0)
+			continue;
+
+		err = fdt_clk_stm32_parse_pll(fdt, subnode, pll);
+		if (err != 0)
+			panic();
+	}
+
+	return 0;
+}
+
+static int fdt_stm32_clk_parse_opp(const void *fdt, int node,
+				   const char *opp_name,
+				   struct stm32_clk_opp_cfg *opp_cfg)
+{
+	int subnode = 0;
+	unsigned int nb_opp = 0;
+	int ret = 0;
+
+	node = fdt_subnode_offset(fdt, node, opp_name);
+	if (node == -FDT_ERR_NOTFOUND)
+		return 0;
+	if (node < 0)
+		return node;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		if (nb_opp >= MAX_OPP) {
+			EMSG("%d MAX opp in %s", MAX_OPP, opp_name);
+			panic();
+		}
+
+		opp_cfg->frq = fdt_read_uint32_default(fdt, subnode,
+						       "hz",
+						       UINT32_MAX);
+
+		opp_cfg->src = fdt_read_uint32_default(fdt, subnode,
+						       "st,clksrc",
+						       UINT32_MAX);
+
+		opp_cfg->div = fdt_read_uint32_default(fdt, subnode,
+						       "st,clkdiv",
+						       UINT32_MAX);
+
+		ret = fdt_clk_stm32_parse_pll(fdt, subnode, &opp_cfg->pll_cfg);
+		if (ret)
+			return ret;
+
+		opp_cfg++;
+		nb_opp++;
+	}
+
+	return 0;
+}
+
+static int fdt_stm32_clk_parse_all_opp(const void *fdt, int node,
+				       struct stm32_clk_platdata *pdata)
+{
+	struct stm32_clk_opp_dt_cfg *opp = pdata->opp;
+	int ret = 0;
+
+	node = fdt_subnode_offset(fdt, node, "st,clk_opp");
+	/* No opp are defined */
+	if (node == -FDT_ERR_NOTFOUND)
+		return 0;
+	if (node < 0)
+		return node;
+
+	ret = fdt_stm32_clk_parse_opp(fdt, node, "st,ck_mpu", opp->mpu_opp);
+	if (ret)
+		return ret;
+
+	ret = fdt_stm32_clk_parse_opp(fdt, node, "st,ck_axi", opp->axi_opp);
+	if (ret)
+		return ret;
+
+	ret = fdt_stm32_clk_parse_opp(fdt, node, "st,ck_mcu", opp->mcu_opp);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int fdt_stm32_clk_parse(const void *fdt, int node,
+			       struct stm32_clk_platdata *pdata)
+{
+	int err = 0;
+
+	err = fdt_stm32_clk_parse_all_oscillator(fdt, node, pdata);
+	if (err != 0)
+		return err;
+
+	err = fdt_stm32_clk_parse_all_pll(fdt, node, pdata);
+	if (err != 0)
+		return err;
+
+	err = fdt_stm32_clk_parse_all_opp(fdt, node, pdata);
+	if (err != 0)
+		return err;
+
+	err = clk_stm32_parse_fdt_by_name(fdt, node, "st,clkdiv", pdata->clkdiv,
+					  &pdata->nclkdiv);
+	if (err != 0)
+		return err;
+
+	err = clk_stm32_parse_fdt_by_name(fdt, node, "st,clksrc", pdata->clksrc,
+					  &pdata->nclksrc);
+	if (err != 0)
+		return err;
+
+	return 0;
+}
+
+/* Clock TREE configuration */
+static int stm32_clk_configure_mux(uint32_t data)
+{
+	int mux = (data & MUX_ID_MASK) >> MUX_ID_SHIFT;
+	int sel = (data & MUX_SEL_MASK) >> MUX_SEL_SHIFT;
+
+	if (mux == MUX_RTC) {
+		/* Mux RTC clock only is selector is valid and RTC not yet
+		 * enabled
+		 */
+		if (sel == 0)
+			return 0;
+
+		if (stm32_gate_is_enabled(GATE_RTCCK))
+			return 0;
+	}
+
+	if (stm32_mux_set_parent(mux, sel))
+		return -1;
+
+	return 0;
+}
+
+static int stm32_clk_configure_clk(struct clk_stm32_priv *priv __maybe_unused,
+				   uint32_t data)
+{
+	int sel = (data & CLK_SEL_MASK) >> CLK_SEL_SHIFT;
+	int enable = (data & CLK_ON_MASK) >> CLK_ON_SHIFT;
+	int clk_id = (data & CLK_ID_MASK) >> CLK_ID_SHIFT;
+
+	if (clk_id == RTC) {
+		uintptr_t address = priv->base + RCC_BDCR;
+
+		if (((io_read32(address) & RCC_BDCR_RTCCKEN) == 0U) || enable) {
+			io_clrsetbits32(address, RCC_BDCR_RTCSRC_MASK,
+					(sel & RCC_SELR_SRC_MASK) <<
+					RCC_BDCR_RTCSRC_SHIFT);
+
+			io_setbits32(address, RCC_BDCR_RTCCKEN);
+		}
+	}
+
+	return 0;
+}
+
+static int stm32_clk_configure_by_addr_val(struct clk_stm32_priv *priv,
+					   uint32_t data)
+{
+	uint32_t addr = data >> CLK_ADDR_SHIFT;
+	uint32_t val = data & CLK_ADDR_VAL_MASK;
+
+	io_setbits32(priv->base + addr, val);
+
+	return 0;
+}
+
+static int stm32_clk_source_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	bool ckper_disabled = false;
+	int ret = 0;
+	size_t i = 0;
+
+	for (i = 0; i < pdata->nclksrc; i++) {
+		uint32_t val = pdata->clksrc[i];
+		uint32_t cmd = 0;
+		uint32_t cmd_data = 0;
+
+		if (val & CMD_ADDR_BIT) {
+			ret = stm32_clk_configure_by_addr_val(priv, val &
+							      ~CMD_ADDR_BIT);
+			if (ret != 0)
+				return ret;
+
+			continue;
+		}
+
+		if (val == (uint32_t)CLK_CKPER_DISABLED) {
+			ckper_disabled = true;
+			continue;
+		}
+
+		cmd = (val & CMD_MASK) >> CMD_SHIFT;
+		cmd_data = val & ~CMD_MASK;
+
+		switch (cmd) {
+		case CMD_MUX:
+			ret = stm32_clk_configure_mux(cmd_data);
+			break;
+
+		case CMD_CLK:
+			ret = stm32_clk_configure_clk(priv, cmd_data);
+			break;
+		default:
+			ret = -1;
+			break;
+		}
+
+		if (ret != 0)
+			return ret;
+	}
+
+	/*
+	 * CKPER is source for some peripheral clocks
+	 * (FMC-NAND / QPSI-NOR) and switching source is allowed
+	 * only if previous clock is still ON
+	 * => deactivate CKPER only after switching clock
+	 */
+	if (ckper_disabled) {
+		ret = stm32_clk_configure_mux(CLK_CKPER_DISABLED);
+		if (ret != 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+#define MAX_HSI_HZ		64000000
+
+static int stm32_clk_hsidiv_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_HSI];
+
+	return stm32_div_set_rate(DIV_HSI, osci->freq, MAX_HSI_HZ);
+}
+
+static TEE_Result stm32_clk_configure_div(uint32_t data)
+{
+	int div_id = (data & DIV_ID_MASK) >> DIV_ID_SHIFT;
+	int div_n = (data & DIV_DIVN_MASK) >> DIV_DIVN_SHIFT;
+
+	return stm32_div_set_value(div_id, div_n);
+}
+
+static int stm32_clk_dividers_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	unsigned int i = 0;
+
+	for (i = 0; i < pdata->nclkdiv; i++) {
+		if (stm32_clk_configure_div(pdata->clkdiv[i]))
+			return -1;
+	}
+
+	return 0;
+}
+
+static unsigned long clk_stm32_pll_get_oscillator_rate(int sel)
+{
+	const int osc[] = { OSC_HSI, OSC_HSE, OSC_CSI };
+
+	assert(sel >= 0 && sel < (int)ARRAY_SIZE(osc));
+
+	return clk_stm32_get_rate_oscillateur(osc[sel]);
+}
+
+static int clk_stm32_pll_compute_cfgr1(const struct stm32_clk_pll *pll,
+				       struct stm32_pll_vco *vco,
+				       uint32_t *value)
+{
+	int sel = (vco->src & MUX_SEL_MASK) >> MUX_SEL_SHIFT;
+	uint32_t divm = vco->div_mn[PLL_CFG_M];
+	uint32_t divn = vco->div_mn[PLL_CFG_N];
+	unsigned long refclk = 0UL;
+
+	refclk = clk_stm32_pll_get_oscillator_rate(sel) / (divm + 1U);
+
+	if ((refclk < (stm32mp1_pll[pll->plltype].refclk_min * 1000000U)) ||
+	    (refclk > (stm32mp1_pll[pll->plltype].refclk_max * 1000000U)))
+		return -1;
+
+	*value = 0;
+
+	if (pll->plltype == PLL_800 && refclk >= 8000000U)
+		*value = 1U << RCC_PLLNCFGR1_IFRGE_SHIFT;
+
+	*value |= (divn << RCC_PLLNCFGR1_DIVN_SHIFT) & RCC_PLLNCFGR1_DIVN_MASK;
+	*value |= (divm << RCC_PLLNCFGR1_DIVM_SHIFT) & RCC_PLLNCFGR1_DIVM_MASK;
+
+	return 0;
+}
+
+static uint32_t  clk_stm32_pll_compute_cfgr2(struct stm32_pll_output *out)
+{
+	uint32_t value = 0;
+
+	value |= (out->output[PLL_CFG_P] << RCC_PLLNCFGR2_DIVP_SHIFT) &
+		 RCC_PLLNCFGR2_DIVP_MASK;
+	value |= (out->output[PLL_CFG_Q] << RCC_PLLNCFGR2_DIVQ_SHIFT) &
+		 RCC_PLLNCFGR2_DIVQ_MASK;
+	value |= (out->output[PLL_CFG_R] << RCC_PLLNCFGR2_DIVR_SHIFT) &
+		 RCC_PLLNCFGR2_DIVR_MASK;
+
+	return value;
+}
+
+/*
+ * Check if PLL1 can be configured on the fly.
+ * @result (-1) => config on the fly is not possible.
+ *         (0)  => config on the fly is possible.
+ *         (+1) => same parameters, no need to reconfigure.
+ * Return value is 0 if no error.
+ */
+static int clk_stm32_is_pll_config_on_the_fly(struct clk_stm32_priv *priv,
+					      const struct stm32_clk_pll *pll,
+					      struct stm32_pll_dt_cfg *pll_conf,
+					      int *result)
+{
+	uintptr_t pll_base = priv->base + pll->reg_pllxcr;
+	struct stm32_pll_vco *vco = &pll_conf->vco;
+	struct stm32_pll_output *out = &pll_conf->output;
+	uint32_t fracr = 0;
+	uint32_t value = 0;
+	int ret = 0;
+	size_t sel = 0;
+
+	ret = clk_stm32_pll_compute_cfgr1(pll, vco, &value);
+	if (ret != 0)
+		return ret;
+
+	sel = (vco->src & MUX_SEL_MASK) >> MUX_SEL_SHIFT;
+	if (sel != stm32_mux_get_parent(pll->mux_id)) {
+		/* Clock source of the PLL is different */
+		*result = -1;
+		return 0;
+	}
+
+	if (io_read32(pll_base + RCC_OFFSET_PLLXCFGR1) != value) {
+		/* Different DIVN/DIVM, can't config on the fly */
+		*result = -1;
+		return 0;
+	}
+
+	*result = 1;
+
+	fracr = vco->frac << RCC_PLLNFRACR_FRACV_SHIFT;
+	fracr |= RCC_PLLNCFGR1_DIVM_MASK;
+	value = clk_stm32_pll_compute_cfgr2(out);
+
+	if ((io_read32(pll_base + RCC_OFFSET_PLLXFRACR) == fracr) &&
+	    (io_read32(pll_base + RCC_OFFSET_PLLXCFGR2) == value)) {
+		/* Same parameters, no need to config */
+		*result = 1;
+	} else {
+		*result = 0;
+	}
+
+	return 0;
+}
+
+static void clk_stm32_pll_config_vco(struct clk_stm32_priv *priv,
+				     const struct stm32_clk_pll *pll,
+				     struct stm32_pll_vco *vco)
+{
+	uintptr_t pll_base = priv->base + pll->reg_pllxcr;
+	uint32_t value = 0;
+
+	if (clk_stm32_pll_compute_cfgr1(pll, vco, &value) != 0) {
+		EMSG("Invalid Vref clock");
+		panic();
+	}
+
+	/* Write N / M / IFREGE fields */
+	io_write32(pll_base + RCC_OFFSET_PLLXCFGR1, value);
+
+	/* Fractional configuration */
+	io_write32(pll_base + RCC_OFFSET_PLLXFRACR, 0);
+
+	/* Frac must be enabled only once its configuration is loaded */
+	io_write32(pll_base + RCC_OFFSET_PLLXFRACR,
+		   vco->frac << RCC_PLLNFRACR_FRACV_SHIFT);
+
+	io_setbits32(pll_base + RCC_OFFSET_PLLXFRACR, RCC_PLLNFRACR_FRACLE);
+}
+
+static void clk_stm32_pll_config_csg(struct clk_stm32_priv *priv,
+				     const struct stm32_clk_pll *pll,
+				     struct stm32_pll_vco *vco)
+{
+	uintptr_t pll_base = priv->base + pll->reg_pllxcr;
+	uint32_t mod_per = 0;
+	uint32_t inc_step = 0;
+	uint32_t sscg_mode = 0;
+	uint32_t value = 0;
+
+	if (!vco->csg_enabled)
+		return;
+
+	mod_per = vco->csg[PLL_CSG_MOD_PER];
+	inc_step = vco->csg[PLL_CSG_INC_STEP];
+	sscg_mode = vco->csg[PLL_CSG_SSCG_MODE];
+
+	value |= (mod_per << RCC_PLLNCSGR_MOD_PER_SHIFT) &
+		 RCC_PLLNCSGR_MOD_PER_MASK;
+	value |= (inc_step << RCC_PLLNCSGR_INC_STEP_SHIFT) &
+		 RCC_PLLNCSGR_INC_STEP_MASK;
+	value |= (sscg_mode << RCC_PLLNCSGR_SSCG_MODE_SHIFT) &
+		 RCC_PLLNCSGR_SSCG_MODE_MASK;
+
+	io_write32(pll_base + RCC_OFFSET_PLLXCSGR, value);
+	io_setbits32(pll_base + RCC_OFFSET_PLLXCR, RCC_PLLNCR_SSCG_CTRL);
+}
+
+static void clk_stm32_pll_config_out(struct clk_stm32_priv *priv,
+				     const struct stm32_clk_pll *pll,
+				     struct stm32_pll_output *out)
+{
+	uintptr_t pll_base = priv->base + pll->reg_pllxcr;
+	uint32_t value = 0;
+
+	value = clk_stm32_pll_compute_cfgr2(out);
+
+	io_write32(pll_base + RCC_OFFSET_PLLXCFGR2, value);
+}
+
+static struct stm32_pll_dt_cfg *clk_stm32_pll_get_pdata(int pll_idx)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct stm32_clk_platdata *pdata = priv->pdata;
+
+	return &pdata->pll[pll_idx];
+}
+
+static int clk_stm32_pll_init_switch_to_hsi_clk_system(int mux_sys)
+{
+	int sel = 0;
+
+	if (mux_sys == -1)
+		return -1;
+
+	/* Make a backup to the current parent */
+	sel = stm32_mux_get_parent(mux_sys);
+
+	/* Switch to HSI */
+	if (stm32_mux_set_parent(mux_sys, 0))
+		return -1;
+
+	return sel;
+}
+
+static uint32_t
+clk_stm32_pll_backup_output_diven(const struct stm32_clk_pll *pll)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	uintptr_t addr = priv->base + pll->reg_pllxcr;
+
+	return io_read32(addr + RCC_OFFSET_PLLXCR) &
+			 (RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
+			  RCC_PLLNCR_DIVREN);
+}
+
+static void clk_stm32_pll_restore_output_diven(const struct stm32_clk_pll *pll,
+					       uint32_t value)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	uintptr_t addr = priv->base + pll->reg_pllxcr;
+	const uint32_t mask = RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
+			      RCC_PLLNCR_DIVREN;
+
+	io_clrsetbits32(addr, mask, value & mask);
+}
+
+static int clk_stm32_pll_init(struct clk_stm32_priv *priv, int pll_idx,
+			      struct stm32_pll_dt_cfg *pll_conf)
+{
+	const struct stm32_clk_pll *pll = clk_stm32_pll_data(pll_idx);
+	int config_on_the_fly = -1;
+	int ret = 0;
+	uint8_t sel = 0;
+	uint32_t save_div_pqr_en = 0;
+	int mux_system[] = { MUX_MPU, MUX_AXI, MUX_MCU, -1 };
+	int mux_sys  = mux_system[pll_idx];
+
+	ret = clk_stm32_is_pll_config_on_the_fly(priv, pll, pll_conf,
+						 &config_on_the_fly);
+	if (ret != 0)
+		return ret;
+
+	/* Backup status of DIV DIVPEN / DIVQEN / DIVREN */
+	save_div_pqr_en = clk_stm32_pll_backup_output_diven(pll);
+
+	if (config_on_the_fly == -1) {
+		/* Make a backup to the current parent and switch to HSI */
+		sel = clk_stm32_pll_init_switch_to_hsi_clk_system(mux_sys);
+
+		/* Stop the PLL before */
+		if (stm32_gate_is_enabled(pll->gate_id))  {
+			io_clrbits32(priv->base + pll->reg_pllxcr,
+				     RCC_PLLNCR_DIVPEN | RCC_PLLNCR_DIVQEN |
+				     RCC_PLLNCR_DIVREN);
+
+			if (stm32_gate_rdy_disable(pll->gate_id))
+				return -1;
+		}
+
+		/* Configure PLLs source */
+		ret = stm32_clk_configure_mux(pll_conf->vco.src);
+		if (ret)
+			return ret;
+
+		clk_stm32_pll_config_vco(priv, pll, &pll_conf->vco);
+	}
+
+	if (config_on_the_fly != 1) {
+		clk_stm32_pll_config_out(priv, pll, &pll_conf->output);
+		clk_stm32_pll_config_csg(priv, pll, &pll_conf->vco);
+	}
+
+	if (!stm32_gate_is_enabled(pll->gate_id)) {
+		if (stm32_gate_rdy_enable(pll->gate_id))
+			return -1;
+
+		clk_stm32_pll_restore_output_diven(pll, save_div_pqr_en);
+	}
+
+	if ((config_on_the_fly == -1) && (mux_sys != -1)) {
+		/* Restore to backup parent */
+		if (stm32_mux_set_parent(mux_sys, sel))
+			return -1;
+	}
+
+	return 0;
+}
+
+static int stm32_clk_pll_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_pll_dt_cfg *pll_conf = NULL;
+	size_t i = 0;
+	const int plls[] = { _PLL1, _PLL3, _PLL4 };
+
+	for (i = 0; i < ARRAY_SIZE(plls); i++) {
+		pll_conf = clk_stm32_pll_get_pdata(plls[i]);
+
+		if (pll_conf->vco.status) {
+			int err = 0;
+
+			err = clk_stm32_pll_init(priv, plls[i], pll_conf);
+			if (err)
+				return err;
+		}
+	}
+
+	return 0;
+}
+
+static int stm32mp1_init_clock_tree(struct clk_stm32_priv *priv,
+				    struct stm32_clk_platdata *pdata)
+{
+	int ret = 0;
+
+	/*
+	 * Switch ON oscillators found in device-tree.
+	 * Note: HSI already ON after BootROM stage.
+	 */
+	stm32_clk_oscillators_enable(priv, pdata);
+
+	ret = stm32_clk_hsidiv_configure(priv);
+	if (ret != 0)
+		panic();
+
+	ret = stm32_clk_dividers_configure(priv);
+	if (ret != 0)
+		panic();
+
+	ret = stm32_clk_pll_configure(priv);
+	if (ret != 0)
+		panic();
+
+	/* Wait LSE ready before to use it */
+	ret = stm32_clk_oscillators_wait_lse_ready(priv, pdata);
+	if (ret != 0)
+		panic();
+
+	/* Configure with expected clock source */
+	ret = stm32_clk_source_configure(priv);
+	if (ret != 0)
+		panic();
+
+	/* Configure LSE CSS after RTC source configuration */
+	ret = stm32_clk_oscillators_lse_set_css(priv, pdata);
+	if (ret != 0)
+		panic();
+
+	/* Software Self-Refresh mode (SSR) during DDR initilialization */
+	io_clrsetbits32(priv->base + RCC_DDRITFCR, RCC_DDRITFCR_DDRCKMOD_MASK,
+			RCC_DDRITFCR_DDRCKMOD_SSR <<
+			RCC_DDRITFCR_DDRCKMOD_SHIFT);
+
+	return 0;
+}
+
+static struct stm32_pll_dt_cfg mp15_pll[_PLL_NB];
+static struct stm32_clk_opp_dt_cfg mp15_clk_opp;
+static struct stm32_osci_dt_cfg mp15_osci[NB_OSC];
+static uint32_t mp15_clksrc[MUX_NB];
+static uint32_t mp15_clkdiv[DIV_NB];
+
+static struct stm32_clk_platdata stm32mp15_clock_pdata = {
+	.osci		= mp15_osci,
+	.nosci		= NB_OSC,
+	.pll		= mp15_pll,
+	.opp		= &mp15_clk_opp,
+	.npll		= _PLL_NB,
+	.clksrc		= mp15_clksrc,
+	.nclksrc	= MUX_NB,
+	.clkdiv		= mp15_clkdiv,
+	.nclkdiv	= DIV_NB,
+};
+
+static struct clk_stm32_priv stm32mp15_clock_data = {
+	.muxes			= parent_mp15,
+	.nb_muxes		= ARRAY_SIZE(parent_mp15),
+	.gates			= gates_mp15,
+	.nb_gates		= ARRAY_SIZE(gates_mp15),
+	.div			= dividers_mp15,
+	.nb_div			= ARRAY_SIZE(dividers_mp15),
+	.pdata			= &stm32mp15_clock_pdata,
+};
+DECLARE_KEEP_PAGER(stm32mp15_clock_data);
+
+static TEE_Result stm32mp1_clock_provider_probe(const void *fdt, int offs,
+						const void *compat_data)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk_stm32_priv *priv = &stm32mp15_clock_data;
+	struct stm32_clk_platdata *pdata = &stm32mp15_clock_pdata;
+	int rc = 0;
+
+	if (compat_data == &non_secure_rcc)
+		disable_rcc_tzen();
+	else
+		enable_rcc_tzen();
+
+	rc = fdt_stm32_clk_parse(fdt, offs, pdata);
+	if (rc) {
+		EMSG("Failed to parse clock node: %d", rc);
+		return TEE_ERROR_GENERIC;
+	}
+
+	res = clk_stm32_init(priv, stm32_rcc_base());
+	if (res)
+		return res;
+
+	rc = stm32mp1_init_clock_tree(priv, pdata);
+	if (rc)
+		return TEE_ERROR_GENERIC;
+
+	res = register_stm32mp1_clocks();
+	if (res) {
+		EMSG("Failed to register clocks: %#"PRIx32, res);
+		panic();
+	}
+
+	res = clk_dt_register_clk_provider(fdt, offs, stm32mp1_clk_dt_get_clk,
+					   NULL);
+	if (res) {
+		EMSG("Failed to register clock provider: %#"PRIx32, res);
+		panic();
+	}
+
+	enable_static_secure_clocks();
+
+	register_pm_core_service_cb(stm32_clock_pm, NULL,
+				    "stm32mp15-clk-service");
+
+	return TEE_SUCCESS;
+}
+
+static const struct dt_device_match stm32mp1_clock_match_table[] = {
+	{  .compatible = "st,stm32mp1-rcc", .compat_data = &non_secure_rcc, },
+	{  .compatible = "st,stm32mp1-rcc-secure", },
+	{ }
+};
+
+DEFINE_DT_DRIVER(stm32mp1_clock_dt_driver) = {
+	.name = "stm32mp1_clock",
+	.type = DT_DRIVER_CLK,
+	.match_table = stm32mp1_clock_match_table,
+	.probe = stm32mp1_clock_provider_probe,
+};
+
 struct soc_stop_context {
 	uint32_t pll3cr;
 	uint32_t pll4cr;
@@ -1724,7 +3064,7 @@ static void restore_mcu_subsys_clocks(void)
 
 	io_write32(rcc_base + RCC_MSSCKSELR, ctx->mssckselr);
 
-	if (stm32mp1_set_clkdiv(ctx->mcudivr, rcc_base + RCC_MCUDIVR)) {
+	if (stm32_div_set_value(DIV_MCU, ctx->mcudivr)) {
 		EMSG("Failed to restore MCUDIVR");
 		panic();
 	}
-- 
2.47.2

